### ISSUE TITLE
Clicking or using a keyboard to open popup doesn't set focus to popup

### ISSUE THREAD

**Issue by acbramley** [Leaflet/Leaflet#2199]
We have very strict accessibility guidelines that we must adhere to. Thankfully leaflet almost solved all our issues we had with Openlayers :) 

There are 2 issues that have been raised by our client's accessibility tester. One being that when a popup is opened (either with the keyboard or mouse) the focus is not set to the popup. This is most important for keyboard users as if a user tabs to a feature, presses enter to open the popup, then tries to get to the popup using the keyboard, they have to tab through ALL other features on the map until they hit the popup. Solution could be to set focus directly to the popup, or make it take the next tabindex?

**Comment by IvanSanchez** [comment]
This has been open for almost two years. @acbramley, do you have any thoughts on the current state of things in 1.0.0-beta2?

**Comment by acbramley** [comment]
Unfortunately I don't, I moved off the project that this work was for a long time ago, I can get in touch with some guys that are working on it still and find out if they still have the issue.

**Comment by techieshark** [comment]
I ran across this issue while trying to figure out how to send focus to a popup for keyboard users for accessibility reasons.

Perhaps someone knows a better way to do this, but for now I just used a hack (getting at internal properties) like this:

```
var clickMarker = function (e) {
      // open popup
      var marker = e.target;
      marker.openPopup();

      // hack for accessibility - no good way to tell popup focus to shift ? - see
      // https://github.com/Leaflet/Leaflet/issues/2199
      var popupNode = e.target._popup._contentNode;
      popupNode.querySelector('a').focus();
};
```

**Comment by skorasaurus** [comment]
I've come across this as well and made a test case: 
http://skorasaurus.github.io/maptemplates/leaflet-test-2199.html and http://bl.ocks.org/skorasaurus/9b2b96967335da91eac7676e815755d2
has the same code. 

how to replicate: 
visit the page 
tab through to a specific marker, press enter to open pop-up 

Expected behavior: 
focus is sent to the marker's pop-up, so use can press enter to access content (a hyperlink) in the pop-up. 

What happens: 
Focus goes to the markers that are east of the marker clicked. (If you open the marker further east, the expected behavior occurs). 

I've tried @techieshark 's workaround but that doesn't work; I haven't tried anything else yet.

**Comment by techieshark** [comment]
Confirming that behavior is as described by @skorasaurus. In gif below you can see I tab through markers, then hit enter on one to open/close popup, but with pop-up open then tabbing takes me to other markers instead of to the link contained in the pop-up. 

Only after all markers have tabbed through do I get to the link content.


![leaflet-tab-issue 2018-12-03 at 16 35 17](https://user-images.githubusercontent.com/1072292/49410446-8573c500-f719-11e8-8a77-eeb080c1142a.gif)

**Comment by Malvoz** [comment]
> There are 2 issues that have been raised by our client's accessibility tester.

A _third_ issue that this would solve is that screen readers currently do not announce the popup's content after it has been opened.

Some have suggested setting `tabindex="0"`, but that would add the popup to the natural tab order which I don't think is necessary (`<dialog>` doesn't recieve focus when opened either). Instead a [negative tabindex](https://webaim.org/techniques/keyboard/tabindex#negative-one) (`tabindex="-1"`) could be set on `.leaflet-popup-content-wrapper` to (only) allow for programmatic focus.

In this case, if outlines are still a concern (https://github.com/Leaflet/Leaflet/issues/2199#issuecomment-29514236) and undesired for the programmatically focused popups something like the following should do:
```css
[tabindex="-1"] {
  outline: 0;
}
```

**Comment by Malvoz** [comment]
Below is a suggestion to also focus the marker when a popup has been closed (it could be treated as a separate issue but is closely related to this):

> maybe focus popup content automatically at popupOpen -> the voice can read it -> give focus back to the marker when popup is closed -> user can tab to next marker.

_Originally posted by @remilev in https://github.com/Leaflet/Leaflet/issues/3210#issuecomment-343735774_
